### PR TITLE
v1.12 Backports 2023-05-03

### PR DESCRIPTION
**Pull request by joamaki**
* [ ] #25046 (@nbusseneau)
 * [x] #24770 (@aditighag)
   *  Applied patch to old filename
 * [x] #25033 (@giorio94)
 * [x] #25109 (@sayboras)
   * Applied patch to old filename
 * [x] #25192 (@kaworu)
   * Only part of the patch applied (same as with v1.13 backport)
 * [ ] #25174 (@joestringer)
 * [x] #25189 (@brb)
 * [x] #25219 (@tommyp1ckles)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
for pr in 25046 24770 25033 25109 25192 25174 25189 25219; do contrib/backporting/set-labels.py $pr done 1.12; done
```
or with
```
make add-labels BRANCH=v1.12 ISSUES=25046,24770,25033,25109,25192,25174,25189,25219
```

### CODE REVIEW

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -1113,6 +1117,12 @@ static __always_inline int nodeport_nat_ipv4_fwd(struct __ctx_buff *ctx)
 	if (ret == NAT_PUNT_TO_STACK)
 		ret = CTX_ACT_OK;
 
+	/* If multiple netdevs process an outgoing packet, then this packets will
+	 * be handled multiple times by the "to-netdev" section. This can lead
+	 * to multiple SNATs. To prevent from that, set the SNAT done flag.
+	 */
+	ctx_snat_done_set(ctx);
```
Comment (brb) [inline bpf/lib/nodeport.h]: Could you s/ctx_nat_done_set/bpf_mark_snat_done/?

File: bpf/lib/nodeport.h [hunk 2]
```diff
@@ -190,6 +190,10 @@ static __always_inline int nodeport_nat_ipv6_fwd(struct __ctx_buff *ctx,
 	      snat_v6_process(ctx, NAT_DIR_EGRESS, &target) : CTX_ACT_OK;
 	if (ret == NAT_PUNT_TO_STACK)
 		ret = CTX_ACT_OK;
+
+	/* See the equivalent v4 path for comment */
+	ctx_snat_done_set(ctx);
```
Comment (brb) [inline bpf/lib/nodeport.h]: Ditto.

**Review by brb — CHANGES_REQUESTED** [review]
Thanks for the backport. Unfortunately, `ctx_mark_snat_done` was renamed only in v1.13.

**Review by kaworu — APPROVED** [review]
Thanks @joamaki my patch LGTM.

**Review by giorio94 — APPROVED** [review]
My commit looks good. Thanks!

File: bpf/lib/nodeport.h (hunk 1 above)
Comment (joamaki) [reply bpf/lib/nodeport.h]: Thanks, amended!

**Review by aditighag — APPROVED** [review]
I checked all the commits, and the backports looks good (including the file name changes). Thanks!

**Review by brb — APPROVED** [review]
:rocket:

**Review by sayboras — APPROVED** [review]
Looks good for my commit, thanks 💯

**Review by tommyp1ckles — APPROVED** [review]

**Comment by joamaki** [comment]
/test-backport-1.12

**Comment by joamaki** [comment]
/test-backport-1.12

**Comment by joamaki** [comment]
/test-1.24-net-next

**Comment by YutaroHayakawa** [comment]
`Chart CI Push` is failing with the known issue, and it should be fixed with the latest main. However, retrying the workflow doesn't reflect the latest workflow file. Since it's not required and it's a bit annoying to spend hours to pass the rest of the tests which are already passing, let me skip `Chart CI Push`.

**Pull request merged**